tap_klaviyo_custom/streams.py

```diff
@@ -1,157 +1,122 @@
 
-        return url
+        # Otherwise use default error checking
+        super().validate_response(response)
 
 
+class ListsStream(KlaviyoStream):
+    """A stream for lists."""
 
-class ListMembersStream(RESTStream):
-    """Define custom stream."""
-    name = "list_members"
-    path = "group/{list_id}/members/all"
-    primary_keys = ["email"]
+    name = "lists"
+    path = "lists"
+    primary_keys = ["list_id"]
     replication_key = None
-    schema_filepath = SCHEMAS_DIR / "list_members.json"
-    #Defining the url_base outside of the class results in an error
-    url_base = 'https://a.klaviyo.com/api/v2/'
+    schema_filepath = SCHEMAS_DIR / "lists.json"
 
-    records_jsonpath = "$.records[*]"
+    @property
+    def partitions(self) -> Optional[List[dict]]:
+        """Get stream partitions."""
+        return [{"list_id": id} for id in self.config["list_ids"]]
 
-    def get_url_params(self, partition: Optional[dict]) -> Dict[str, Any]:
-        """Return a dictionary of values to be used in URL parameterization."""
-        params = {}
-        params.update({"api_key": self.config['api_key']})
-        return params
-
-    def prepare_request(
-        self, context: Optional[dict], next_page_token: Optional[Any], list_id: Optional[str]
-    ) -> requests.PreparedRequest:
-        """Prepare a request object.
-
-        If partitioning is supported, the `context` object will contain the partition
-        definitions. Pagination information can be parsed from `next_page_token` if
-        `next_page_token` is not None.
-        """
-        http_method = self.rest_method
-        url: str = self.get_url(context, list_id)
-        params: dict = self.get_url_params(context)
-        request_data = self.prepare_request_payload(context, next_page_token)
-        headers = self.http_headers
-
-        if next_page_token != None:
-            params['marker'] = next_page_token
-
-        request = cast(
-            requests.PreparedRequest,
-            self.requests_session.prepare_request(
-                requests.Request(
-                    method=http_method,
-                    url=url,
-                    params=params,
-                    headers=headers,
-                    json=request_data,
-                )
-            ),
-        )
-        # sleep timer to avoid Klaviyo API rate limit errors
-        time.sleep(1)
-        return request
+    def get_next_page_token(
+        self, response: requests.Response, previous_token: Optional[Any]
+    ) -> Any:
+        """Return token identifying next page or None if all records have been read."""
+        # since we're using partitions, there's only ever one page per partition
+        return None
 
-    def get_url(self, context: Optional[dict], list_id: Optional[str]) -> str:
-        """Return a URL, optionally targeted to a specific partition or context.
+    def post_process(self, row: dict, context: Optional[dict] = None) -> Optional[dict]:
```

Comment:
I'm a little confused as to what's happening in this ListsStream and specifically in the post_process function. The API call made in my version of the ListsStream produces a dict of all of the Klaviyo list names and IDs. Is that still happening somewhere here? I assume it's happening automatically, but that also confuses me as to why you have to append the list ID in the return statement.

Reply:
I changed ListsStream to pull only the lists in the config instead of all the lists. On one hand, it seems like we might as well pull all the lists. However, doing so would kinda mess up the parent-child stream model, and there isn't any reason to pull the lists that aren't in the config.

The reason it's a little tricky to keep the existing behavior with child streams is the SDK doesn't have a way to conditionally generate child streams, so you can't say "get all the lists, and if the list id is one of these, also get the members." Instead, this uses partitions to say "get each one of these lists by id" and then child context to say "for each of these lists, also get the members."

Reply:
Ahh, that makes a lot more sense now. Does the partition function limit the lists API call to only pull in list IDs defined in the config?

Reply:
Yeah. Basically `partition` works kinda like child streams but without a parent. So the partition creates a new ListsStream for each list in the config. Each ListStream pulls one stream by ID, and then triggers a child ListMembersStream